include/zenoh/api/shm/provider/shm_provider_backend.hxx

```diff
@@ -0,0 +1,60 @@
+//
+// Copyright (c) 2023 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+#pragma once
+
+#include "../../../detail/interop.hxx"
+#include "../../base.hxx"
+#include "chunk.hxx"
+#include "types.hxx"
+
+namespace zenoh {
+
+class CppShmProviderBackendIface {
+   public:
+    virtual ChunkAllocResult alloc(const MemoryLayout &layout) = 0;
+    virtual void free(const ChunkDescriptor &chunk) = 0;
+    virtual size_t defragment() = 0;
+    virtual size_t available() const = 0;
+    virtual void layout_for(MemoryLayout &layout) = 0;
+    virtual ~CppShmProviderBackendIface() = default;
+};
+
+class CppShmProviderBackend : public CppShmProviderBackendIface {};
+
+class CppShmProviderBackendThreadsafe : public CppShmProviderBackend {};
+
+// Ensure that function pointers are defined with extern C linkage
+namespace shm::provider_backend::closures {
+extern "C" {
+inline void del(void *context) { delete static_cast<CppShmProviderBackend *>(context); }
```

Comment:
Can you rename functions in extern "C" block (and prefix with _z ?) to avoid symbol name collisions